### PR TITLE
 Make Bucket using builder pattern

### PR DESCRIPTION
**Pull request by sinhaashish**
@balamurugana @harshavardhana  
Can you  please validate this approach,

### CODE REVIEW

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.minio;
+
+public class MakeBucketArgs {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:19]: 1. You need to have `newBuilder()` which returns `MakeBucketArgs.Builder`
2. `Builder()` doesn't create another `Builder`
3. Do not assume default region is `us-east-1`. The default region is the one passed to `MinioClient` or `us-east-1`. Here you don't have much control.
4. Refer https://square.github.io/okhttp/3.x/okhttp/okhttp3/Headers.html for clarity.

File: api/src/main/java/io/minio/MinioClient.java [hunk 2]
```diff
@@ -3359,6 +3359,65 @@ public void makeBucket(String bucketName, String region, boolean objectLock)
...
+   * @throws NoSuchAlgorithmException thrown to indicate missing of MD5 or SHA-256 digest library.
+   * @throws RegionConflictException thrown to indicate passed region conflict with default region.
+   * @throws XmlParserException thrown to indicate XML parsing error.
+   */
+  public void makeBucket(MakeBucketArgs makebucketArgs)
+      throws InvalidBucketNameException, RegionConflictException, InsufficientDataException,
+          InternalException, InvalidResponseException, InvalidKeyException,
+          NoSuchAlgorithmException, XmlParserException, ErrorResponseException, IOException {
+
+    checkBucketName(makebucketArgs.bucketName());
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java:3388]: 0. Just have `public void makeBucket(MakeBucketArgs args)` and check `args` is not `null`.
1. All such validations need to go into `MakeBucketArgs.Builder` and throw `IllegalArgumentException` on error.
2. All existing `makeBucket()` should use this `makeBucket()` and they can be deprecated.
3. As all existing `makeBucket()` are deprecated, we would need to move examples/functional tests to use this `makeBucket()`

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  public static final class Builder {
+    private String bucketName;
+    private String region;
+    private boolean objectLock;
+
+    public Builder bucket(String bucketName) throws IllegalArgumentException {
+
+      if (bucketName == null) {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:54]: 1. `BucketName` and `ObjectName` are common type across `minio-py`. You could make these classes to ensure their correctness.
2. Any argument builder may accept `String bucketName` and/or `String objectName` and `BucketName` and `ObjectName` objects will be created internally.
3. My suggestion is to go through https://github.com/square/okhttp/blob/okhttp_3.14.x/okhttp/src/main/java/okhttp3/Headers.java as a reference for builder support.

File: api/src/main/java/io/minio/BucketName.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  public BucketName() {}
+
+  public String bucketName() {
+    return bucketName;
+  }
+
+  public BucketName(String bucketName) {
+
+    if (bucketName == null) {
+      throw new IllegalArgumentException(NULL_STRING + "null bucket name");
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketName.java:32]: You could just have `null bucket name`

File: api/src/main/java/io/minio/BucketName.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.minio;
+
+public class BucketName {
+  private String bucketName;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketName.java:20]: 1. `name` is good enough here.
2. Have `public String toString()` returning `name` to stringify this class.

File: api/src/main/java/io/minio/BucketName.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  private static final String NULL_STRING = "(null)";
+
+  public BucketName() {}
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketName.java:23]: No arg constructor is incorrect.

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  public BucketName bucket() {
+    return this.bucket;
+  }
+
+  public boolean objectLock() {
+    return objectLock;
+  }
+
+  public static final class Builder {
+    private String bucketName;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:50]: you would need to hold `BucketName` object here, not string.

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    private String region;
+    private boolean objectLock;
+
+    public Builder bucket(String bucketName) throws IllegalArgumentException {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:54]: We would need one more method `public Builder bucket(BucketName name)` and you would need to check for `null` value.

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+      this.bucketName = bucketName;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:55]: create `BucketName` object and keep it.

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.minio;
+
+public final class MakeBucketArgs {
+  private final String bucketName;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:20]: Keep `BucketName` object.

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  private final String region;
+  private final boolean objectLock;
+  private final BucketName bucket;
+  private static final String NULL_STRING = "(null)";
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:24]: Also remove all unwanted constants and member variables.

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      return this;
+    }
+
+    public Builder objectLock(boolean objectLock) {
+      this.objectLock = objectLock;
+      return this;
+    }
+
+    public MakeBucketArgs build() {
+      return new MakeBucketArgs(this);
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:70]: you would check and throw `IllegalArgumentException` if `bucketName` object is `null`

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+  private final boolean objectLock;
+  private final BucketName bucket;
+  private static final String NULL_STRING = "(null)";
+
+  MakeBucketArgs(Builder builder) {
+    this.bucket = new BucketName(builder.bucketName);
+    this.bucketName = builder.bucketName;
+    this.region = builder.region;
+    this.objectLock = builder.objectLock;
+  }
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:31]: We would need `newBuilder()` method to create `Builder`.

File: api/src/main/java/io/minio/MinioClient.java [hunk 14]
```diff
@@ -3263,12 +3263,17 @@ public boolean bucketExists(String bucketName)
...
       throws ErrorResponseException, IllegalArgumentException, InsufficientDataException,
           InternalException, InvalidBucketNameException, InvalidKeyException,
           InvalidResponseException, IOException, NoSuchAlgorithmException, RegionConflictException,
           XmlParserException {
-    this.makeBucket(bucketName, null, false);
+
+    MakeBucketArgs.Builder makeBucketArgs = new MakeBucketArgs.Builder();
+    makeBucketArgs.bucket(bucketName);
+    MakeBucketArgs args = makeBucketArgs.build();
+    this.makeBucket(args);
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java:3276]: You could use `this.makeBucket(MakeBucketArgs.newBuilder().bucket(bucketName).build())` directly everywhere.

File: api/src/main/java/io/minio/MinioClient.java [hunk 15]
```diff
@@ -3324,14 +3333,56 @@ public void makeBucket(String bucketName, String region)
...
+  public void makeBucket(MakeBucketArgs args)
+      throws InvalidBucketNameException, RegionConflictException, InsufficientDataException,
+          InternalException, InvalidResponseException, InvalidKeyException,
+          NoSuchAlgorithmException, XmlParserException, ErrorResponseException, IOException {
+
+    if ((args == null)) {
+      throw new IllegalArgumentException("null value is not allowed in arguments");
+    }
+
+    String region = null;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java:3379]: initialize with `US_EAST_1` than `null` and remove `else` part.

File: api/src/main/java/io/minio/MinioClient.java [hunk 16]
```diff
@@ -3324,14 +3333,56 @@ public void makeBucket(String bucketName, String region)
...
+    if (args.region() != null && !args.region().isEmpty()) {
+      region = args.region();
+    } else if (this.region != null && !this.region.isEmpty()) {
       region = this.region;
+    } else {
+      region = US_EAST_1;
     }
 
     // If constructor already sets a region, check if it is equal to region param if provided
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java:3388]: I am thinking to get rid of constructor driven `region` setting and have `BucketName` object coming with `region` value itself.  That could be a separate PR after this PR merged.

File: docs/API.md [hunk 17]
```diff
@@ -646,55 +646,22 @@ for (Result<Item> result : results) {
...
-|:---------------|:---------|:--------------------|
-| ``bucketName`` | _String_ | Name of the bucket. |
+| Parameter      | Type             | Description                |
+|:---------------|:-----------------|:---------------------------|
+| ``args      `` | _MakeBucketArgs_ | Arguments to create bucket |
 
 __Example__
 ```java
-minioClient.makeBucket("my-bucketname");
+MakeBucketArgs args = MakeBucketArgs.newBuilder().bucket("mybucket").region("us-east-1").objectLock(true).build();
```
Comment (balamurugana) [inline docs/API.md:661]: 1. This example code should be spotless checked.
2. Provide multiple examples like
```
// Create bucket with default region.
minioClient.makeBucket(MakeBucketArgs.newBuilder().bucket("my-bucketname").build());

// Create bucket with specific region.
....

// Create object-lock enabled bucket with specific region.
...
```

File: api/src/main/java/io/minio/BucketName.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.minio;
+
+public class BucketName {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketName.java:19]: Have this class as `Bucket`

Comment (sinhaashish) [reply api/src/main/java/io/minio/BucketName.java]: Renamed it to `BucketTemplate` as Bucket class was already present .

File: api/src/main/java/io/minio/BucketTemplate.java [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+package io.minio;
+
+public class BucketTemplate {
+  private String name;
+
+  public String name() {
+    return name;
+  }
+
+  public BucketTemplate(String bucketName) {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketTemplate.java:26]: Use `name` instead of `bucketName`

File: api/src/main/java/io/minio/BucketTemplate.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.minio;
+
+public class BucketTemplate {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketTemplate.java:19]: Have name as `Bucket` and use fully qualified name in `public List<Bucket> listBuckets()`.

File: api/src/main/java/io/minio/BucketTemplate.java [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+          "bucket name does not follow Amazon S3 standards. For more information refer "
+              + "http://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html";
+      throw new IllegalArgumentException(bucketName + " : " + msg);
+    }
+    this.name = bucketName;
+  }
+
+  @Override
+  public String toString() {
+    return "BucketTemplate{" + "name='" + name + '\'' + '}';
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketTemplate.java:58]: Just `return this.name;`

File: api/src/main/java/io/minio/MinioClient.java [hunk 22]
```diff
@@ -3293,12 +3295,29 @@ public void makeBucket(String bucketName)
...
+    this.makeBucket(
+        MakeBucketArgs.newBuilder().bucket("my-bucketname").region("us-east-1").build());
+
+    // Create object-lock enabled bucket with specific region.
+    this.makeBucket(
+        MakeBucketArgs.newBuilder()
+            .bucket("my-bucketname")
+            .region("us-east-1")
+            .objectLock(true)
+            .build());
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java:3318]: Remove above three examples.

File: api/src/main/java/io/minio/MinioClient.java [hunk 23]
```diff
@@ -3324,13 +3343,55 @@ public void makeBucket(String bucketName, String region)
...
+   * @throws NoSuchAlgorithmException thrown to indicate missing of MD5 or SHA-256 digest library.
+   * @throws RegionConflictException thrown to indicate passed region conflict with default region.
+   * @throws XmlParserException thrown to indicate XML parsing error.
+   */
+  public void makeBucket(MakeBucketArgs args)
+      throws InvalidBucketNameException, RegionConflictException, InsufficientDataException,
+          InternalException, InvalidResponseException, InvalidKeyException,
+          NoSuchAlgorithmException, XmlParserException, ErrorResponseException, IOException {
+
+    if ((args == null)) {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java:3387]: Why double bracket?

File: docs/API.md [hunk 24]
```diff
@@ -646,55 +646,34 @@ for (Result<Item> result : results) {
...
+    minioClient.makeBucket(
+        MakeBucketArgs.newBuilder().bucket("my-bucketname").region("us-east-1").build());
+
+    // Create object-lock enabled bucket with specific region.
+    minioClient.makeBucket(
+        MakeBucketArgs.newBuilder()
+            .bucket("my-bucketname")
+            .region("us-east-1")
+            .objectLock(true)
+            .build());
```
Comment (balamurugana) [inline docs/API.md:674]: Remove leading spaces.

File: api/src/main/java/io/minio/MinioClient.java [hunk 25]
```diff
@@ -3323,10 +3315,6 @@ public void makeBucket(String bucketName, String region)
   /**
    * Creates a bucket with object lock feature enabled.
    *
-   * <pre>Example:{@code
-   * minioClient.makeBucket("my-bucketname", "eu-west-2", true);
-   * }</pre>
-   *
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java:3310]: You could retain these examples till respective methods are removed.

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 26]
```diff
@@ -44,18 +44,18 @@ public static Builder newBuilder() {
   }
 
   public static final class Builder {
-    private BucketTemplate bucket;
+    private Bucket bucket;
     private String region;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:48]: I think we need to keep region inside `Bucket` object. `Builder` would create appropriate `Bucket` object.

File: api/src/main/java/io/minio/MinioClient.java [hunk 27]
```diff
@@ -3324,13 +3331,55 @@ public void makeBucket(String bucketName, String region)
...
+            .objectLock(objectLock)
+            .build());
+  }
+
+  /**
+   * Creates a bucket makeBucketArgs.
+   *
+   * <pre>Example:{@code
+   * minioClient.makeBucket(MakeBucketArgs makebucketArgs);
+   * }</pre>
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java:3354]: You would need sync up description and examples with APIs.md

File: docs/API.md [hunk 28]
```diff
@@ -646,55 +646,34 @@ for (Result<Item> result : results) {
...
-Creates a bucket with default region.
+Creates a bucket with given region and object lock feature enabled.
 
 __Parameters__
-| Parameter      | Type     | Description         |
-|:---------------|:---------|:--------------------|
-| ``bucketName`` | _String_ | Name of the bucket. |
+| Parameter      | Type             | Description                |
+|:---------------|:-----------------|:---------------------------|
+| ``args      `` | _MakeBucketArgs_ | Arguments to create bucket |
```
Comment (balamurugana) [inline docs/API.md:657]: 1. Remove whitespace in ``args``.
2. You would need to add link to `[MakeBucketArgs]`.

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.minio;
+
+public final class MakeBucketArgs {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:19]: Please note that this class expect `objectLock` is required to each and every `<API>Args` class in future. We do not copy/paste this class, instead we need a base class `BucketArgs` and other classes inherit it.

APIs accepting object name as argument require different style and we will be fixing the design in one of such PR.

@kannappanr Any intermediate release in the middle will break backward compatibility. Its required to have close attentions on these PRs.

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 30]
```diff
@@ -16,19 +16,15 @@
...
 
-public final class MakeBucketArgs {
-  private final Bucket bucket;
+public final class MakeBucketArgs extends BucketArgs {
   private final boolean objectLock;
 
   MakeBucketArgs(Builder builder) {
-    this.bucket = builder.bucket;
+    this.name = builder.name;
+    this.region = builder.region;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:24]: `BucketArgs` already has these. You would need to make them `protected` in `BucketArgs`

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 31]
```diff
@@ -37,34 +33,16 @@ public static Builder newBuilder() {
...
+  public static final class Builder extends BucketArgs.BucketArgsBuilder {
     private boolean objectLock;
 
     public Builder bucket(String name) {
-      if (this.bucket == null) {
-        this.bucket = new Bucket(name, this.region);
-      } else {
-        this.bucket = new Bucket(name, this.bucket.region());
-      }
+      this.name = name;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:40]: You don't need these methods.

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 32]
```diff
@@ -37,34 +33,16 @@ public static Builder newBuilder() {
...
-      }
-      return this;
-    }
-
-    public Builder bucket(Bucket bucket) throws IllegalArgumentException {
-      if (bucket == null) {
-        throw new IllegalArgumentException("null value is not allowed for Bucket");
-      }
-      this.bucket = bucket;
+      this.region = region;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:45]: Same as above.

**Comment by sinhaashish** [comment]
PR https://github.com/minio/minio-java/pull/927 addresses this .
